File: src/frontend/src/types/placeCategories.ts
/**
 * High-level place categories for traveler-friendly filtering
 */
export type PlaceCategory =
  | 'food_dining'
  | 'coffee_shops'
  | 'groceries'
  | 'attractions'
  | 'shopping'
  | 'nature_parks'
  | 'healthcare'
  | 'services'
  | 'transportation'
  | 'nightlife';

/**
 * Mapping of high-level categories to Google Places API type identifiers
 */
export const CATEGORY_MAPPINGS: Record<PlaceCategory, string[]> = {
  food_dining: ['restaurant', 'cafe', 'bar', 'bakery', 'meal_takeaway'],
  coffee_shops: ['cafe', 'coffee_shop'],
  groceries: ['supermarket', 'grocery_store', 'convenience_store'],
  attractions: [
    'tourist_attraction',
    'museum',
    'art_gallery',
    'landmark',
    'point_of_interest',
    'aquarium',
    'zoo',
    'amusement_park',
    'beach',
  ],
  shopping: ['shopping_mall', 'store', 'clothing_store', 'department_store'],
  nature_parks: ['park', 'natural_feature', 'campground', 'beach'],
  healthcare: ['pharmacy', 'hospital', 'doctor'],
  services: ['atm', 'bank', 'post_office', 'laundry'],
  transportation: [
    'gas_station',
    'parking',
    'transit_station',
    'bus_station',
    'train_station',
  ],
  nightlife: ['night_club', 'bar', 'casino', 'movie_theater'],
};

/**
 * Human-readable labels for categories
 */
export function getCategoryLabel(category: PlaceCategory): string {
  const labels: Record<PlaceCategory, string> = {
    food_dining: 'Food & Dining',
    coffee_shops: 'Coffee Shops',
    groceries: 'Groceries & Essentials',
    attractions: 'Attractions & Culture',
    shopping: 'Shopping',
    nature_parks: 'Nature & Parks',
    healthcare: 'Healthcare',
    services: 'Services',
    transportation: 'Transportation',
    nightlife: 'Nightlife & Entertainment',
  };
  return labels[category];
}

/**
 * Material icon identifiers for categories
 */
export function getCategoryIcon(category: PlaceCategory): string {
  const icons: Record<PlaceCategory, string> = {
    food_dining: 'restaurant',
    coffee_shops: 'local_cafe',
    groceries: 'local_grocery_store',
    attractions: 'attractions',
    shopping: 'shopping_bag',
    nature_parks: 'park',
    healthcare: 'local_pharmacy',
    services: 'room_service',
    transportation: 'directions_bus',
    nightlife: 'nightlife',
  };
  return icons[category];
}

/**
 * Description text for categories
 */
export function getCategoryDescription(category: PlaceCategory): string {
  const descriptions: Record<PlaceCategory, string> = {
    food_dining: 'Restaurants, cafes, bars, and bakeries',
    coffee_shops: 'Coffee shops and cafes',
    groceries: 'Supermarkets, grocery stores, and convenience stores',
    attractions: 'Museums, landmarks, beaches, tourist attractions, and points of interest',
    shopping: 'Shopping malls, stores, and retail locations',
    nature_parks: 'Parks, beaches, natural features, and campgrounds',
    healthcare: 'Pharmacies, hospitals, and medical services',
    services: 'ATMs, banks, post offices, and laundry services',
    transportation: 'Gas stations, parking, and transit stations',
    nightlife: 'Night clubs, bars, casinos, and entertainment venues',
  };
  return descriptions[category];
}

/**
 * All available categories for iteration
 */
export const ALL_CATEGORIES: PlaceCategory[] = [
  'food_dining',
  'coffee_shops',
  'groceries',
  'attractions',
  'shopping',
  'nature_parks',
  'healthcare',
  'services',
  'transportation',
  'nightlife',
];

File: src/backend/src/MapInsights.Core/Places/PlaceCategories.cs
namespace MapInsights.Core.Places;

/// <summary>
/// High-level place categories for traveler-friendly filtering.
/// Each category maps to multiple Google Places API type identifiers.
/// </summary>
public enum PlaceCategory
{
    FoodDining,
    CoffeeShops,
    Groceries,
    Attractions,
    Shopping,
    NatureParks,
    Healthcare,
    Services,
    Transportation,
    Nightlife
}

/// <summary>
/// Maps high-level categories to Google Places API type identifiers
/// </summary>
public static class CategoryMappings
{
    private static readonly Dictionary<PlaceCategory, string[]> _mappings = new()
    {
        [PlaceCategory.FoodDining] = ["restaurant", "cafe", "bar", "bakery", "meal_takeaway"],
        [PlaceCategory.CoffeeShops] = ["cafe", "coffee_shop"],
        [PlaceCategory.Groceries] = ["supermarket", "grocery_store", "convenience_store"],
        [PlaceCategory.Attractions] = ["tourist_attraction", "museum", "art_gallery", "landmark", "point_of_interest", "aquarium", "zoo", "amusement_park", "beach"],
        [PlaceCategory.Shopping] = ["shopping_mall", "store", "clothing_store", "department_store"],
        [PlaceCategory.NatureParks] = ["park", "natural_feature", "campground", "beach"],
        [PlaceCategory.Healthcare] = ["pharmacy", "hospital", "doctor"],
        [PlaceCategory.Services] = ["atm", "bank", "post_office", "laundry"],
        [PlaceCategory.Transportation] = ["gas_station", "parking", "transit_station", "bus_station", "train_station"],
        [PlaceCategory.Nightlife] = ["night_club", "bar", "casino", "movie_theater"]
    };

    /// <summary>
    /// Gets Google Places API types for a single category
    /// </summary>
    public static string[] GetGooglePlaceTypes(PlaceCategory category)
    {
        return _mappings.TryGetValue(category, out var types) ? types : Array.Empty<string>();
    }

    /// <summary>
    /// Gets all Google Places API types for multiple categories, flattened and deduplicated
    /// </summary>
    public static List<string> GetAllGooglePlaceTypes(IEnumerable<PlaceCategory> categories)
    {
        return categories
            .SelectMany(GetGooglePlaceTypes)
            .Distinct()
            .ToList();
    }

    /// <summary>
    /// Parses a category string to PlaceCategory enum (case-insensitive, handles underscore_case)
    /// </summary>
    public static PlaceCategory? ParseCategory(string categoryString)
    {
        if (string.IsNullOrWhiteSpace(categoryString))
            return null;

        // Try direct parsing first (handles PascalCase)
        if (Enum.TryParse<PlaceCategory>(categoryString, ignoreCase: true, out var category))
            return category;

        // Convert underscore_case to PascalCase (e.g., "food_dining" -> "FoodDining")
        var pascalCase = string.Join("", categoryString.Split('_')
            .Select(word => char.ToUpperInvariant(word[0]) + word.Substring(1).ToLowerInvariant()));

        return Enum.TryParse<PlaceCategory>(pascalCase, ignoreCase: false, out category)
            ? category
            : null;
    }
}
